smart_open/http.py

```diff
@@ -75,32 +84,42 @@ def detach(self):
         """Unsupported."""
         raise io.UnsupportedOperation
 
-    def read(self, size=None):
+    def read(self, size=-1):
         """
         Mimics the read call to a filehandle object.
         """
-        logger.debug('read: %r', locals())
-        if size is None:
-            return self.response.raw.read()
+        logger.debug("reading with size: %d", size)
+        if self.response is None:
+            return ''
+
+        if size == 0:
+            return ''
+        elif size < 0:
```

Comment:
Unless there's a good reason to nest conditionals here, I think rewriting them as un-nested is better.

```python
elif size < 0 and not self._read_buffer:
    retval = self.response.raw.read()
elif size < 0:
    retval = self._read_buffer + self.response.raw.read()
    self._read_buffer = b''
```